modules/Evse15118D20/charger/ISO15118_chargerImpl.hpp

```diff
@@ -0,0 +1,99 @@
+    // ev@8ea32d28-373f-4c90-ae5e-b4fcc74e2a61:v1
+
+protected:
+    // command handler functions (virtual)
+    virtual void handle_setup(types::iso15118_charger::EVSEID& evse_id,
+                              std::vector<types::iso15118_charger::EnergyTransferMode>& supported_energy_transfer_modes,
+                              types::iso15118_charger::SaeJ2847BidiMode& sae_j2847_mode, bool& debug_mode,
+                              bool& bidirectional) override;
+    virtual void handle_set_charging_parameters(types::iso15118_charger::SetupPhysicalValues& physical_values) override;
+    virtual void handle_session_setup(std::vector<types::iso15118_charger::PaymentOption>& payment_options,
+                                      bool& supported_certificate_service) override;
+    virtual void
+    handle_certificate_response(types::iso15118_charger::ResponseExiStreamStatus& exi_stream_status) override;
+    virtual void handle_authorization_response(types::authorization::AuthorizationStatus& authorization_status,
+                                               types::authorization::CertificateStatus& certificate_status) override;
+    virtual void handle_ac_contactor_closed(bool& status) override;
+    virtual void handle_dlink_ready(bool& value) override;
+    virtual void handle_cable_check_finished(bool& status) override;
+    virtual void handle_receipt_is_required(bool& receipt_required) override;
+    virtual void handle_stop_charging(bool& stop) override;
+    virtual void handle_update_ac_max_current(double& max_current) override;
+    virtual void handle_update_dc_maximum_limits(types::iso15118_charger::DcEvseMaximumLimits& maximum_limits) override;
+    virtual void handle_update_dc_minimum_limits(types::iso15118_charger::DcEvseMinimumLimits& minimum_limits) override;
+    virtual void handle_update_isolation_status(types::iso15118_charger::IsolationStatus& isolation_status) override;
+    virtual void handle_update_dc_present_values(
+        types::iso15118_charger::DcEvsePresentVoltageCurrent& present_voltage_current) override;
+    virtual void handle_update_meter_info(types::powermeter::Powermeter& powermeter) override;
+    virtual void handle_send_error(types::iso15118_charger::EvseError& error) override;
+    virtual void handle_reset_error() override;
+
+    // ev@d2d1847a-7b88-41dd-ad07-92785f06f5c4:v1
+    // insert your protected definitions here
+    // ev@d2d1847a-7b88-41dd-ad07-92785f06f5c4:v1
+
+private:
+    const Everest::PtrContainer<Evse15118D20>& mod;
+    const Conf& config;
+
+    virtual void init() override;
+    virtual void ready() override;
+
+    // ev@3370e4dd-95f4-47a9-aaec-ea76f34a66c9:v1
+    iso15118::session::feedback::Callbacks create_callbacks();
+
+    std::unique_ptr<iso15118::TbdController> controller;
+
+    static constexpr auto WAIT_FOR_SETUP_DONE_MS{std::chrono::milliseconds(200)};
+
+    iso15118::d20::EvseSetupConfig setup_config;
+    std::uint8_t setup_state{0};
```

Comment:
`setup_state` itself can't be an enum (except you're defining all product states), but probably a [std::bitset](https://en.cppreference.com/w/cpp/utility/bitset). The `EnumFlags` could be a good choice, but it comes with some drawbacks:
- it enforces the use of std::atomic, which should be optional (zero overhead principle), although this is required here
- the enum is required to have a `last` field
- it doesn't have an `all_set` function
- you can't get it's current value (you can only test single bits, but this only within one atomic operation, so you can't in principle tell at all if multiple bits are set at the same time)

Concerning the misuse of `=` and `|=`: this is debatable, because the user might like to be able to set one specific bit at once, without clearing/resetting beforehand.

Reply:
std::bitset would be fine if available, EnumFlags isn't suitable as you have stated - it was an example of how you can get bit flag operation with enum type safety - perhaps a starting point.
A new class taking on those ideas could also address the locking issue (or use atomic flags ...)

Reply:
Changed `setup_state` to a `std::bitset`